spec.html

```diff
@@ -30110,19 +30110,54 @@ <h1>Function.prototype.toString ( )</h1>
-          1. If Type(_func_) is Object and IsCallable(_func_) is *true*, return an implementation-defined String source code representation of _func_. The representation must have the syntax of a |NativeFunction|.
-          1. Throw a *TypeError* exception.
+          1. Else if _func_ is a <emu-xref href="#sec-built-in-function-objects">built-in function object</emu-xref>, then
+            1. Let _sourceText_ be an implementation-defined sequence of Unicode code points representing _func_.
+          1. Else if Type(_func_) is Object and IsCallable(_func_) is *true*, then
+            1. Let _sourceText_ be an implementation-defined sequence of Unicode code points representing _func_.
+          1. Else,
+            1. Throw a *TypeError* exception.
+          1. Assert: ParseText(_sourceText_, |NativeFunction|) is a Parse Node.
+          1. If _func_ has an [[InitialName]] internal slot and _func_.[[InitialName]] is a String, then
+            1. Assert: NativeFunctionName(_sourceText_) is _func_.[[InitialName]].
+          1. Return CodePointsToString(_sourceText_).
         </emu-alg>
 
         <emu-grammar type="definition">
           NativeFunction :
-            `function` NativeFunctionAccessor? PropertyName[~Yield, ~Await]? `(` FormalParameters[~Yield, ~Await] `)` `{` `[` `native` `code` `]` `}`
+            `function` NativeFunctionName? `(` FormalParameters[~Yield, ~Await] `)` `{` `[` `native` `code` `]` `}`
+
+          NativeFunctionName :
+            NativeFunctionAccessor? PropertyName[~Yield, ~Await]
 
           NativeFunctionAccessor :
             `get`
             `set`
         </emu-grammar>
+
+        <emu-clause id="sec-nativefunctionname" type="abstract operation">
+          <h1>
+            NativeFunctionName (
+              _sourceText_: a sequence of Unicode code points,
+            ): a String
+          </h1>
+          <dl class="header">
+            <dt>description</dt>
+            <dd>It returns the name of a built-in function object represented by source text.</dd>
+          </dl>
+          <emu-alg>
+            1. Let _function_ be ParseText(_sourceText_, |NativeFunction|).
+            1. If _name_ Contains |NativeFunctionName| is *false*, return *""*.
+            1. Let _name_ be the |NativeFunctionName| of _function_.
+            1. Let _propertyNameNode_ be the |PropertyName| of _name_.
+            1. Let _propertyName_ be the result of evaluating _propertyNameNode_.
+            1. Assert: Type(_propertyName_) is String.
+            1. NOTE: The |PropertyName| production in general may be a |ComputedPropertyName| evaluating to a Symbol (e.g., `[Symbol()]`), but this is not the case for any built-in function that is directly accessible to ECMAScript code.
```

Comment:
Where is this restriction coming from? I don't think this is true.

Reply:
This code is only executed in fulfillment of an assertion about [[InitialName]], which is only set in SetFunctionName via CreateBuiltinFunction—and whose requirements are somewhat scattered. Every explicit use of CreateBuiltinFunction specifies _name_ as a string, and every implicit use is covered by the combination of [CreateIntrinsics](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-createintrinsics) ("_All values that are built-in <emu-xref href="#function-object" id="_ref_3587"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#function-object">function objects</a></emu-xref> are created by performing <emu-xref aoid="CreateBuiltinFunction" id="_ref_3588"><a href="https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-createbuiltinfunction">CreateBuiltinFunction</a></emu-xref>(<var>steps</var>, <var>length</var>, <var>name</var>, <var>slots</var>, <var>realmRec</var>, <var>prototype</var>) where… <var>name</var> is the initial value of the function's <emu-val>"name"</emu-val> property_") and [ECMAScript Standard Built-in Objects](https://tc39.es/ecma262/multipage/ecmascript-standard-built-in-objects.html#sec-ecmascript-standard-built-in-objects):
> Each built-in function defined in this specification is created by calling the <emu-xref aoid="CreateBuiltinFunction" id="_ref_8096"><a href="https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-createbuiltinfunction">CreateBuiltinFunction</a></emu-xref> abstract operation (<emu-xref href="#sec-createbuiltinfunction" id="_ref_511"><a href="https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-createbuiltinfunction">10.3.3</a></emu-xref>). The values of the <var>length</var> and <var>name</var> parameters are the initial values of the <emu-val>"length"</emu-val> and <emu-val>"name"</emu-val> properties as discussed below. The values of the <var>prefix</var> parameter are similarly discussed below.
> …
> Every built-in <emu-xref href="#function-object" id="_ref_8104"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#function-object">function object</a></emu-xref>, including <emu-xref href="#constructor" id="_ref_8105"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#constructor">constructors</a></emu-xref>, has a <emu-val>"name"</emu-val> property whose value is a String.
> …
> The value of the <emu-val>"name"</emu-val> property is explicitly specified for each built-in functions whose <emu-xref href="#sec-object-type" id="_ref_8108"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-object-type">property key</a></emu-xref> is a Symbol value. [<em>for example, [Symbol.prototype [ @@toPrimitive ]](https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-symbol.prototype-@@toprimitive) and [get RegExp [ @@species ]](https://tc39.es/ecma262/multipage/text-processing.html#sec-get-regexp-@@species)</em>]